src/Swift.Bindings/src/Marshaler/BoundGenericsHandler.cs

```diff
@@ -0,0 +1,146 @@
+        return !s_structGenerics.Contains(swiftTypeName);
+    }
+
+    /// <summary>
+    /// Translates the Swift generic type of the given property declaration into a C# type name.
+    /// </summary>
+    /// <param name="propertyDecl">The property declaration.</param>
+    /// <returns>The C# type name with generic parameters.</returns>
+    /// <exception cref="NotSupportedException">Thrown when the property is not bound generic.</exception>
+    public string TranslateToCSharpName(PropertyDecl propertyDecl)
+    {
+        if (!IsBoundGeneric(propertyDecl))
+            throw new NotSupportedException(
+                $"Attempted to translate to C# name for a non-bound generic property {propertyDecl.Name}");
+        var namedTypeSpec = (NamedTypeSpec)propertyDecl.SwiftTypeSpec;
+        return TranslateToCSharpName(namedTypeSpec);
+    }
+
+    /// <summary>
+    /// Translates the Swift generic type of the given argument declaration into a C# type name.
+    /// </summary>
+    /// <param name="argumentDecl">The argument declaration.</param>
+    /// <returns>The C# type name with generic parameters.</returns>
+    /// <exception cref="NotSupportedException">Thrown when the argument is not bound generic.</exception>
+    public string TranslateToCSharpName(ArgumentDecl argumentDecl)
+    {
+        if (!IsBoundGeneric(argumentDecl))
+            throw new NotSupportedException(
+                $"Attempted to translate to C# name for a non-bound generic argument {argumentDecl.Name}");
+        var namedTypeSpec = (NamedTypeSpec)argumentDecl.SwiftTypeSpec;
+        return TranslateToCSharpName(namedTypeSpec);
+    }
+
+    /// <summary>
+    /// Helper method to convert a Swift <see cref="NamedTypeSpec"/> into its corresponding C# type name.
+    /// </summary>
+    /// <param name="namedTypeSpec">The named type specification.</param>
+    /// <returns>The C# type name string.</returns>
+    /// <exception cref="NotSupportedException">
+    /// Thrown when any generic parameter is not a named type specification.
+    /// </exception>
+    private string TranslateToCSharpName(NamedTypeSpec namedTypeSpec)
+    {
+        List<string> translatedGenericParameters = new();
+        foreach (var genericParameter in namedTypeSpec.GenericParameters)
+        {
+            if (genericParameter is not NamedTypeSpec namedGenericParameter)
+                throw new NotSupportedException(
+                    $"Generic parameter {genericParameter} is not a named type spec");
+            translatedGenericParameters.Add(TranslateToCSharpName(namedGenericParameter));
+        }
+
+        var typeReference = _typeDatabase.GetTypeRecordOrAnyType(namedTypeSpec); // TODO: consider throwing an exception instead
+        return typeReference.CSTypeIdentifier +
+               (translatedGenericParameters.Count > 0
+                    ? $"<{string.Join(", ", translatedGenericParameters)}>"
+                    : "");
```

Comment:
Curious - when would this happen? It's a bound generic type but it has no generic parameters? Is that the case `T` alone (so type parameter of the enclosing type?)

Reply:
This will happen at last level of recursion So If I have e.g. `Foo<Bar<Int>>` this will be true for Int

Reply:
In that case I will add another comment which I was thinking originally about: Currently this implementes Swift->C# name translation only for generic entities. Would it make sense to make a more general version which can work on non-generic entities as well?
The "bottom of the recursion" level in generic case is the same as the non-generic case anyway.

Reply:
>Would it make sense to make a more general version which can work on non-generic entities as well?
Absolutely, we have it on the backlog already. https://github.com/dotnet/runtimelab/issues/2889. Implementing that will affect a lot of code so I would like to bring it as a separate change. 

Renamed the function so it should be clear what it does